### PR TITLE
Improve dotnet nuget push to Azure DevOps Artifacts documentation

### LINKED ISSUE #2787: Improve dotnet nuget push to Azure DevOps Artifacts documentation
Current document doesn't cover how to push Azure DevOps properly with `dotnet nuget`.
This is a follow up on https://github.com/NuGet/Home/issues/7792

### PR DESCRIPTION
**Pull request by erdembayar**
Fixes: https://github.com/NuGet/docs.microsoft.com-nuget/issues/2787

Description: Improve dotnet nuget push to Azure DevOps Artifacts documentation: https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-nuget-push
Add example for pushing nupkg to Azdo Artifacts.

### CODE REVIEW

File: docs/core/tools/dotnet-nuget-push.md [hunk 1]
```diff
@@ -55,6 +55,7 @@ The command pushes an existing package. It doesn't create a package. To create a
 - **`-k|--api-key <API_KEY>`**
 
   The API key for the server.
+  - Azure DevOps Artifacts feeds doesn't accept PAT(personal access tokens) passed directly in cli as `API_KEY`. For dev box, you need have the [cred provider](https://github.com/microsoft/artifacts-credprovider) installed and that would work for both push and download (see examples section). For CI, you need use the [NuGet Authenticate](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate?view=azure-devops) task with [NuGet service connection](https://docs.microsoft.com/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml#nuget-service-connection) for pushing to Azure DevOps Artifacts.
```
Comment (nkolev92) [inline docs/core/tools/dotnet-nuget-push.md:58]: I think this is slightly out of place here. 

IMO the description should match what the tooling provides. 

Pushing to Azure DevOps can be its own section.

File: docs/core/tools/dotnet-nuget-push.md [hunk 2]
```diff
@@ -148,5 +149,25 @@ The command pushes an existing package. It doesn't create a package. To create a
   ```dotnetcli
   dotnet nuget push "*.nupkg" -s c:\mydir
   ```
+  
+- In the example below pushing "Foo" version "5.0.2" to Azure DevOps Artifacts from dev box, here AZ is just a placeholder for ApiKey, this prevents authentication fail prematurely, in order to authentication to work you need to install [cred provider](https://github.com/microsoft/artifacts-credprovider). Below command trigger open Cred Provider window if authentication is necessary, it's suitable for pushing from dev box, but not for CI.
```
Comment (nkolev92) [inline docs/core/tools/dotnet-nuget-push.md:153]: Similarly to above, I think we should have a dedicated section for azure devops.

Comment (nkolev92) [reply docs/core/tools/dotnet-nuget-push.md:153]: Feels like we should be referencing an azure devops page with instructions for this instead.

Does Azure Devops have documentation on how to handle pushing at all?

**Review by nkolev92 — COMMENTED** [review]
I was mostly focusing on the structure first right now. 

Hoping @tdykstra can provide some guidance on this. 

I haven't reviewed the specific text yet.

File: docs/core/tools/dotnet-nuget-push.md (hunk 2 above)
Comment (zivkan) [reply docs/core/tools/dotnet-nuget-push.md:153]: AzDO does have this page on publishing packages: https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?toc=%2Fazure%2Fdevops%2Fartifacts%2Ftoc.json&bc=%2Fazure%2Fdevops%2Fartifacts%2Fbreadcrumb%2Ftoc.json&view=azure-devops&tabs=yaml


I don't feel 100% comfortable putting feed specific instructions in our push docs that's generic for all feeds. We certainly do have a lot of Azure Artifacts customers, and it is a Microsoft service. And if customers come to this help page, but don't find Azure Artifacts' docs page, then instructions here can help them be successful.

On the other hand, GitHub and nuget.org and are also owned by Microsoft. Should we put instructions for those NuGet servers in this doc too?  How about non-Microsoft servers like MyGet, GitLabs, TeamCity, Artifactory, and who knows how many more?  I might be just too idealistic, but I don't feel like server specific information belong in the generic tool docs.

Comment (zivkan) [reply docs/core/tools/dotnet-nuget-push.md:153]: Also, isn't the .NET SDK and CLI supposed to be intentionally platform agnostic?  To help dispel the myth that .NET is for Microsoft platforms only. I'm not sure that documenting special instructions for Azure Artifacts only helps maintain that independence.

File: docs/core/tools/dotnet-nuget-push.md (hunk 1 above)
Comment (jeffkl) [inline docs/core/tools/dotnet-nuget-push.md:58]: The abbreviation should be in parentheses and then you can use the abbreviation later now that its been spelled out once.

```suggestion
  - Azure Artifacts feeds do not accept a personal access token (PAT) specified by this command-line argument. When using a local development environment, you must have the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed which provides authentication when pushing or downloading packages (see examples section). For hosted builds that interfact with Azure Artifacts feeds, you must use the [NuGet Authenticate](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate) task.
```

File: docs/core/tools/dotnet-nuget-push.md (hunk 2 above)
Comment (jeffkl) [inline docs/core/tools/dotnet-nuget-push.md:153]:
```suggestion
- In the example below, pushing "MyPackage" version "5.0.2" to an Azure Artifacts feed from a local development environment, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) to be installed to properly authenticate.
```

File: docs/core/tools/dotnet-nuget-push.md [hunk 3]
```diff
@@ -148,5 +149,25 @@ The command pushes an existing package. It doesn't create a package. To create a
...
+
+  ```dotnetcli
+  dotnet nuget push Foo.5.0.2.nupkg --source https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json -k AZ
```
Comment (jeffkl) [inline docs/core/tools/dotnet-nuget-push.md:156]:
```suggestion
  dotnet nuget push MyPackage.5.0.2.nupkg --source https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json -k AZ
```

File: docs/core/tools/dotnet-nuget-push.md [hunk 4]
```diff
@@ -148,5 +149,25 @@ The command pushes an existing package. It doesn't create a package. To create a
...
+  ```
+  
+- In the example below "Foo" version "1.1.5" nupkg to Azure DevOps Artifacts from CI, here AZ is just a placeholder for ApiKey, this prevents authentication fail prematurely. You need to setup [NuGet Authenticate task](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate?view=azure-devops) with [NuGet service connection](https://docs.microsoft.com/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml#nuget-service-connection) for authenticate with external Azure DevOps Artifacts server.
```
Comment (jeffkl) [inline docs/core/tools/dotnet-nuget-push.md:159]:
```suggestion
- In the example below, pushing "MyPackage" version "1.1.5" to an Azure Artifacts feed from a hosted build, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires your pipeline to use the [NuGet Authenticate task](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate?view=azure-devops).
```

File: docs/core/tools/dotnet-nuget-push.md [hunk 5]
```diff
@@ -148,5 +149,25 @@ The command pushes an existing package. It doesn't create a package. To create a
...
+
+  ```bashcli
+    - task: NuGetAuthenticate@1
+      inputs:
+        nuGetServiceConnections: MyServiceConnection_ExternalServer
+        
+    - bash: |
+        dotnet build foo/foo.csproj -c Release
```
Comment (jeffkl) [inline docs/core/tools/dotnet-nuget-push.md:167]:
```suggestion
        dotnet build mypackage/mypackage.csproj -c Release
```

File: docs/core/tools/dotnet-nuget-push.md [hunk 6]
```diff
@@ -148,5 +149,25 @@ The command pushes an existing package. It doesn't create a package. To create a
...
+        dotnet pack foo/foo.csproj /property:PackageVersion=1.1.5 -o nupkgs -c Release
```
Comment (jeffkl) [inline docs/core/tools/dotnet-nuget-push.md:168]:
```suggestion
        dotnet pack mypackage/mypackage.csproj /property:PackageVersion=1.1.5 -o nupkgs -c Release
```

File: docs/core/tools/dotnet-nuget-push.md [hunk 7]
```diff
@@ -148,5 +149,25 @@ The command pushes an existing package. It doesn't create a package. To create a
...
+        dotnet nuget push nupkgs/foo.1.1.5.nupkg --source https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json -k AZ
```
Comment (jeffkl) [inline docs/core/tools/dotnet-nuget-push.md:169]:
```suggestion
        dotnet nuget push nupkgs/mypackage.1.1.5.nupkg --source https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json -k AZ
```

**Review by jeffkl — COMMENTED** [review]
Submitted some wording suggestions.  I know this doc already contains references to `foo` but I don't think we should use that reference in public facing documentation.

File: docs/core/tools/dotnet-nuget-push.md (hunk 2 above)
Comment (erdembayar) [reply docs/core/tools/dotnet-nuget-push.md:153]: > AzDO does have this page on publishing packages: https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?toc=%2Fazure%2Fdevops%2Fartifacts%2Ftoc.json&bc=%2Fazure%2Fdevops%2Fartifacts%2Fbreadcrumb%2Ftoc.json&view=azure-devops&tabs=yaml
> 
This is opposite direction (Azdo -> xxxx) of what I added here. We can't use Azdo task to push nuget to Azdo from local or Github actions.

> On the other hand, GitHub and nuget.org and are also owned by Microsoft. Should we put instructions for those NuGet servers in this doc too? How about non-Microsoft servers like MyGet, GitLabs, TeamCity, Artifactory, and who knows how many more? I might be just too idealistic, but I don't feel like server specific information belong in the generic tool docs.

Yes, you concern is valid. Still most like customer come to this doc page due to search engines so having this info relieve customer frustration. So far, MyGet, GPR (I manually tested them) and most nuget server implemenet uses `apikey`, but Azdo ignores actual value of `apikey` passed, still requires passing something (in this case `AZ`). Considering quite many customer upvoted on https://github.com/NuGet/Home/issues/7792 I believe we need to document it, I'm not aware of any other package server has a quirk like Azdo.

Comment (nkolev92) [reply docs/core/tools/dotnet-nuget-push.md:153]: > I believe we need to document it, I'm not aware of any other package server has an apikey quirk like Azdo.

I agree that we should document it, but it doesn't have to be a section in the push description page.

Comment (zivkan) [reply docs/core/tools/dotnet-nuget-push.md:153]: > This is opposite direction (Azdo -> xxxx) of what I added here.

I copied the wrong URL, but if you looked at the table of contents on the left, it was the article immediately underneath the one I linked: [Publish a NuGet package using the command line](https://docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops). To me, that is exactly what you're documenting. How to publish a nuget package using the CLI, to an Azure Artifacts feed.

Comment (erdembayar) [reply docs/core/tools/dotnet-nuget-push.md]: Created link to example in Azure Artifacts examples.

**Review by nkolev92 — APPROVED** [review]

**Review by IEvangelist — APPROVED** [review]

**Pull request merged**